compiler/rustc_parse/src/lexer/mod.rs

```diff
@@ -54,7 +54,7 @@ pub(crate) fn parse_token_trees<'a>(
         start_pos = start_pos + BytePos::from_usize(shebang_len);
     }
 
-    let cursor = Cursor::new(src);
+    let cursor = Cursor::new(src, sess.edition.rust_2021());
```

Comment:
This doesn't handle tokens correctly whose span has a different edition. E.g. the following UI test should compile & the assertion should hold when run but due to `sess.edition.rust_2021()` it leads to `error: `c".."` literals are experimental` being emitted (and if you enabled the feature inside `user.rs`, the assertion would fail with 1 ≠ 2).

`auxiliary/count.rs`:

```rs
// force-host
// edition:2018
// no-prefer-dynamic
#![crate_type = "proc-macro"]
extern crate proc_macro;
use proc_macro::TokenStream;
use std::str::FromStr;

#[proc_macro]
pub fn number_of_tokens(_: TokenStream) -> TokenStream {
    TokenStream::from_str("c\"\"").unwrap().into_iter().count().to_string().parse().unwrap()
}
```

`user.rs`:

```rs
// run-pass
// edition:2021
// aux-build:count.rs
extern crate count;

fn main() { assert_eq!(count::number_of_tokens!(), 2); }
```

If you instead wrote the code below (which I'm not suggesting you do), then the test case above would compile. This solution is likely incorrect as well though. **petrochenkov** can probably answer that.

```suggestion
    let span = override_span.unwrap_or_else(|| Span::with_root_ctxt(start_pos, start_pos));
    let cursor = Cursor::new(src, span.ctxt().outer_expn_data().edition.rust_2021());
```

Note that my example is basically stolen from [tests/ui/rust-2021/reserved-prefixes-via-macro.rs](https://github.com/rust-lang/rust/blob/master/tests/ui/rust-2021/reserved-prefixes-via-macro.rs).

Reply:
Looks like `override_span` never changes during the life of `Cursor` and `StringReader`, so this should work too.